Scroll corner painting ignores color-scheme when overflow: scroll

lukewarlow opened the pull request:
#### 4498c483beb97701ef5ca9d7ee2050b86a8d924a
<pre>
Scroll corner painting ignores color-scheme when overflow: scroll
<a href="https://bugs.webkit.org/show_bug.cgi?id=258482">https://bugs.webkit.org/show_bug.cgi?id=258482</a>

Reviewed by Aditya Keerthi.

Update the local appearance value before painting Mac scroll corner.

This means that the color-scheme is correctly accounted for when painting the scroll corner for both frame views and scrollable areas.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::paintScrollCorner):
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::paintScrollCorner):

Canonical link: <a href="https://commits.webkit.org/265687@main">https://commits.webkit.org/265687@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10ba7fe98b47e2513c4b2000d888be004f24dcea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11659 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12226 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13301 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11104 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11678 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14247 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11839 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13973 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11823 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12667 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9887 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13724 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9959 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10583 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17717 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11034 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10737 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13914 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11136 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/9185 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10313 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2796 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14593 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10993 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -1476,8 +1480,14 @@ void RenderLayerScrollableArea::paintScrollCorner(GraphicsContext& context, cons
+#if PLATFORM(MAC)
+        // Keep this in sync with ScrollAnimatorMac's effectiveAppearanceForScrollerImp:.
+        LocalDefaultSystemAppearance localAppearance(useDarkAppearance());
+#endif
```
If possible, we should try to avoid adding platform-specific code in a platform-agnostic file.

In this case, I think we can:
1. Move this logic in to `ScrollbarThemeMac::paintScrollCorner`. Name the `ScrollableArea` parameter, and read `area.useDarkAppearanceForScrollbars()`.
2. Remove similar logic from `LocalFrameView::paintScrollCorner`.

If that doesn't work, we can move ahead with this solution.

lukewarlow replied:
Taking a look at that now can't see why it wouldn't work but will test it

pxlcoder commented on Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
```diff
@@ -582,11 +583,14 @@ static void paintScrollbar(Scrollbar& scrollbar, GraphicsContext& context)
         return;
 
+    // Keep this in sync with ScrollAnimatorMac's effectiveAppearanceForScrollerImp:.
+    LocalDefaultSystemAppearance localAppearance(area.useDarkAppearance());
```
This should be `useDarkAppearanceForScrollbars()`.

lukewarlow replied:
Apologies I didn't see that on the type I looked at but I was looking at the wrong type.

lukewarlow replied:
Fixed. Sorry for that.

lukewarlow committed:
Scroll corner painting ignores color-scheme when overflow: scroll
https://bugs.webkit.org/show_bug.cgi?id=258482

Reviewed by Aditya Keerthi.

Update the local appearance value before painting Mac scroll corner.

This means that the color-scheme is correctly accounted for when painting the scroll corner for both frame views and scrollable areas.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::paintScrollCorner):
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::paintScrollCorner):

Canonical link: https://commits.webkit.org/265687@main
